### PR TITLE
Update scheduler table tests to use subtest

### LINKED ISSUE #63267: scheduler: Update tests to use subtest
Go 1.7 added the [subtest feature](https://golang.org/doc/go1.7#testing) which can make table-driven tests much easier to run and debug. Many table-driven tests in `pkg/scheduler` and `test/integration/scheduler` are not using this feature.

As an example, https://github.com/kubernetes/kubernetes/pull/56940 converts some tests to use a sub-benchmark.

Further reading: [Using Subtests and Sub-benchmarks](https://blog.golang.org/subtests)

/kind cleanup
/sig scheduling

### PR DESCRIPTION
**Pull request by xchapter7x**
**What this PR does / why we need it**: Update scheduler's unit table tests to use subtest

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #63267

**Special notes for your reviewer**:
```
Integration tests and benchmark refactor will be in a separate PR to keep the footprint of change 
smaller.

Side note: there are a number of inconsistencies across the subpackages in the scheduler regarding how the tables are implemented. Implementations use maps, struct elements of several different names, etc. Perhaps another refactor issue to unify the approach would be beneficial?
I tried aligning some of the implementations in naming and approach, but likely a separate effort and discussion would be best. thoughts?

```

**Release note**:

```release-note
This PR will leverage subtests on the existing table tests for the scheduler units.
Some refactoring of error/status messages and functions to align with new approach.

```

### CODE REVIEW

**Comment by resouer** [comment]
Thanks, I will take a review after CI is happy

File: pkg/scheduler/scheduler_test.go [hunk 1]
```diff
@@ -185,71 +185,104 @@ func TestScheduler(t *testing.T) {
...
-					return fakeBinder{func(b *v1.Binding) error {
-						gotBinding = b
-						return item.injectBindError
-					}}
-				},
-				PodConditionUpdater: fakePodConditionUpdater{},
-				Error: func(p *v1.Pod, err error) {
-					gotPod = p
-					gotError = err
+		name := fmt.Sprintf("%v:expecterror=%v:injecterror=%v:eventreason=%v", i, item.expectError, item.injectBindError, item.eventReason)
```
Comment (resouer) [inline pkg/scheduler/scheduler_test.go]: Please give a meaningful but short name here, the current name is too long.

**Review by resouer — COMMENTED** [review]
One nits, and pleas squash your commit into one. Thanks!

File: pkg/scheduler/scheduler_test.go (hunk 1 above)
Comment (xchapter7x) [reply pkg/scheduler/scheduler_test.go]: 👍

**Comment by xchapter7x** [comment]
hi @resouer thanks for the feedback.

Regarding your comment on squashing the commits. Is it preferred to keep squashing new commits while in the `wip` status, or were you suggesting I squash the commits before i remove the `wip` status?

**Comment by resouer** [comment]
You can do squash whenever you like :)

**Comment by xchapter7x** [comment]
hi @resouer ,

This PR should be in a good spot for an initial review. Let me know your thoughts and feedback. Thanks :)

-John

File: pkg/scheduler/schedulercache/node_info_test.go [hunk 2]
```diff
@@ -132,13 +145,15 @@ func TestResourceClone(t *testing.T) {
...
 
-	for _, test := range tests {
-		r := test.resource.Clone()
-		// Modify the field to check if the result is a clone of the origin one.
-		test.resource.MilliCPU += 1000
-		if !reflect.DeepEqual(test.expected, r) {
-			t.Errorf("expected: %#v, got: %#v", test.expected, r)
-		}
+	for i, test := range tests {
+		t.Run(fmt.Sprintf("%v/%v", i, test.name), func(t *testing.T) {
```
Comment (misterikkit) [inline pkg/scheduler/schedulercache/node_info_test.go]: Why include the index in the test name when the names are already unique?

File: pkg/scheduler/scheduler_test.go [hunk 3]
```diff
@@ -144,7 +144,7 @@ func TestScheduler(t *testing.T) {
 	errB := errors.New("binder")
 	testNode := v1.Node{ObjectMeta: metav1.ObjectMeta{Name: "machine1", UID: types.UID("machine1")}}
 
-	table := []struct {
+	table := map[string]struct {
```
Comment (misterikkit) [inline pkg/scheduler/scheduler_test.go]: Why change this to a map when the name string is a struct member everywhere else?

File: pkg/scheduler/algorithm/priorities/interpod_affinity_test.go [hunk 4]
```diff
@@ -295,7 +295,7 @@ func TestInterPodAffinityPriority(t *testing.T) {
 				{ObjectMeta: metav1.ObjectMeta{Name: "machine3", Labels: labelAzAz1}},
 			},
 			expectedList: []schedulerapi.HostPriority{{Host: "machine1", Score: schedulerapi.MaxPriority}, {Host: "machine2", Score: 0}, {Host: "machine3", Score: 0}},
-			test: "Affinity: pod that matches topology key & pods in nodes will get high score comparing to others" +
+			name: "Affinity: pod that matches topology key & pods in nodes will get high score comparing to others" +
```
Comment (misterikkit) [inline pkg/scheduler/algorithm/priorities/interpod_affinity_test.go]: I know you tried not to modify existing test names, but I think "Affinity:" is redundant in these test names. If you like, we can do test name cleanup in a separate PR (after this one)

File: pkg/scheduler/algorithm/priorities/node_label_test.go [hunk 5]
```diff
@@ -47,7 +47,7 @@ func TestNewNodeLabelPriority(t *testing.T) {
 			expectedList: []schedulerapi.HostPriority{{Host: "machine1", Score: 0}, {Host: "machine2", Score: 0}, {Host: "machine3", Score: 0}},
 			label:        "baz",
 			presence:     true,
-			test:         "no match found, presence true",
+			name:         "no match found, presence true",
```
Comment (misterikkit) [inline pkg/scheduler/algorithm/priorities/node_label_test.go]: on the subject of cleaning up subtest names, replacing ', ' with '/' in these tests would be nice.

**Review by misterikkit — COMMENTED** [review]
Thanks for doing this work. If you run into merge conflicts, it might be easier to do one PR per package.

**Comment by misterikkit** [comment]
After discussion, I think cleaning up test names is best done in a separate PR.

**Comment by xchapter7x** [comment]
hey @misterikkit , 

as per our discussion i've broken up this PR by package. The new PRs can be found here:
https://github.com/kubernetes/kubernetes/pull/63666
https://github.com/kubernetes/kubernetes/pull/63665
https://github.com/kubernetes/kubernetes/pull/63664
https://github.com/kubernetes/kubernetes/pull/63663
https://github.com/kubernetes/kubernetes/pull/63662
https://github.com/kubernetes/kubernetes/pull/63661
https://github.com/kubernetes/kubernetes/pull/63660
https://github.com/kubernetes/kubernetes/pull/63659
https://github.com/kubernetes/kubernetes/pull/63658

these PRs also contain the discussed changes.

should i close out this PR in favor of the others? Let me know how to proceed. Thanks.

- John

**Comment by xchapter7x** [comment]
closing this in favor of the split up PRs.

@resouer or @misterikkit , 
mind setting those new PRs to `/ok-to-test` ?
thanks in advance.

-John

**Comment by k82cn** [comment]
closed? how do we trace the progress? e.g. how many PR are merged, how many PRs are pending.
